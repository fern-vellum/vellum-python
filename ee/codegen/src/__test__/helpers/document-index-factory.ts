export const mockDocumentIndexFactory = () => ({
  id: "b6ef8809-346e-469c-beed-2e5c4e9844c5",
  name: "my-sweet-document",
});
