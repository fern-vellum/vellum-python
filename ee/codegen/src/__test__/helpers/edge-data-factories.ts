import { WorkflowEdge, WorkflowNode } from "src/types/vellum";

const getSourceHandleId = (
  node: WorkflowNode | [WorkflowNode, string]
): string => {
  if (Array.isArray(node)) {
    const [actualNode, portName] = node;

    if (actualNode.type === "GENERIC") {
      const id = actualNode.ports.find((port) => port.name === portName)?.id;
      if (!id) {
        throw new Error(`Port not found: ${portName}`);
      }

      return id;
    }

    if (actualNode.type === "CONDITIONAL") {
      const id = actualNode.data.conditions[Number(portName)]?.sourceHandleId;
      if (!id) {
        throw new Error("Conditional node has no source handle id");
      }

      return id;
    }

    return getSourceHandleId(actualNode);
  }

  if (node.type === "GENERIC") {
    const port = node.ports[0];
    if (!port) {
      throw new Error("Generic node has no ports");
    }

    return port.id;
  }

  if (
    node.type == "TERMINAL" ||
    node.type === "NOTE" ||
    node.type === "ERROR"
  ) {
    throw new Error(`${node.type} nodes have no source handle id`);
  }

  if (node.type === "CONDITIONAL") {
    throw new Error("Conditional nodes must include a port name");
  }

  return node.data.sourceHandleId;
};

const getTargetHandleId = (
  node: WorkflowNode | [WorkflowNode, number]
): string => {
  if (Array.isArray(node)) {
    const [actualNode, targetIndex] = node;
    if (actualNode.type === "MERGE") {
      const id = actualNode.data.targetHandles[targetIndex]?.id;
      if (!id) {
        throw new Error(`Target handle not found: ${targetIndex}`);
      }

      return id;
    }

    return getTargetHandleId(actualNode);
  }

  if (node.type === "GENERIC") {
    return node.trigger.id;
  }

  if (node.type === "MERGE") {
    throw new Error("Merge nodes must include target handle index");
  }

  if (node.type === "ENTRYPOINT" || node.type === "NOTE") {
    throw new Error(`${node.type} nodes have no target handle id`);
  }

  return node.data.targetHandleId;
};

export type EdgeFactoryNodePair = [
  WorkflowNode | [WorkflowNode, string],
  WorkflowNode | [WorkflowNode, number]
];

export function edgesFactory(nodePairs: EdgeFactoryNodePair[]): WorkflowEdge[] {
  return nodePairs.map(([sourceNode, targetNode], index) => ({
    id: `edge-${index + 1}`,
    type: "DEFAULT",
    sourceNodeId: Array.isArray(sourceNode) ? sourceNode[0].id : sourceNode.id,
    sourceHandleId: getSourceHandleId(sourceNode),
    targetNodeId: Array.isArray(targetNode) ? targetNode[0].id : targetNode.id,
    targetHandleId: getTargetHandleId(targetNode),
  }));
}
