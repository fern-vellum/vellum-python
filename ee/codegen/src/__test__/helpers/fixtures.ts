import { readdirSync } from "fs";
import { basename, join, relative } from "path";

const fixtureDir = join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "tests",
  "codegen_integration",
  "fixtures"
);

const allFixtures = readdirSync(fixtureDir);

function getFixtures(
  excludeFixtures: Set<string> | string[] = new Set(),
  includeFixtures: Set<string> | string[] = new Set()
): string[] {
  const excludeSet = Array.isArray(excludeFixtures)
    ? new Set(excludeFixtures)
    : excludeFixtures;
  const includeSet = Array.isArray(includeFixtures)
    ? new Set(includeFixtures)
    : includeFixtures;

  return allFixtures.filter((f) => {
    const fileName = basename(f);

    return (
      (!excludeSet.size || !excludeSet.has(fileName)) &&
      (!includeSet.size || includeSet.has(fileName))
    );
  });
}

function getFixturePaths(root: string) {
  const displayFile = join(
    fixtureDir,
    root,
    "display_data",
    `${basename(root)}.json`
  );
  const codeDir = join(fixtureDir, root, "code");

  return { displayFile, codeDir };
}

export function getFixturesForProjectTest(
  {
    excludeFixtures,
    includeFixtures,
  }: {
    excludeFixtures?: Set<string> | string[];
    includeFixtures?: Set<string> | string[];
  } = { excludeFixtures: new Set(), includeFixtures: new Set() }
): {
  fixtureName: string;
  displayFile: string;
  codeDir: string;
}[] {
  return getFixtures(excludeFixtures, includeFixtures).map((root) => ({
    fixtureName: root,
    ...getFixturePaths(root),
  }));
}

export function getAllFilesInDir(
  dirPath: string,
  ignoreRegexes?: RegExp[]
): { [relativePath: string]: string } {
  const files: { [relativePath: string]: string } = {};

  function traverseDir(currentPath: string) {
    const entries = readdirSync(currentPath, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name);

      if (entry.isDirectory()) {
        traverseDir(fullPath);
      } else if (entry.isFile()) {
        const relativePath = relative(dirPath, fullPath);

        if (
          ignoreRegexes &&
          ignoreRegexes.some((regex) => regex.test(relativePath))
        ) {
          continue;
        }

        files[relativePath] = fullPath;
      }
    }
  }

  traverseDir(dirPath);
  return files;
}
