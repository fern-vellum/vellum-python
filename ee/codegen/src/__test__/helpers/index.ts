export { nodeContextFactory } from "./node-context-factory";
export { workflowContextFactory } from "./workflow-context-factory";
