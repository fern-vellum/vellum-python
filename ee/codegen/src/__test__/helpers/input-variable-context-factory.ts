import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { InputVariableContext } from "src/context/input-variable-context";

export function inputVariableContextFactory({
  inputVariableData,
  workflowContext,
}: {
  inputVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): InputVariableContext {
  return new InputVariableContext({
    inputVariableData,
    workflowContext,
  });
}
