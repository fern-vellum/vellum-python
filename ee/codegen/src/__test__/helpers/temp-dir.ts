import os from "os";
import path from "path";

export function makeTempDir(base?: string): string {
  const randomSuffix = Math.random().toString(36).substring(7);
  return path.join(os.tmpdir(), `${base ?? "test-directory"}-${randomSuffix}`);
}
