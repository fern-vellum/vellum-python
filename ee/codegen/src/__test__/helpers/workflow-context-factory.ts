import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowLabel,
  workflowClassName,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowLabel || "Workflow",
    workflowLabel: workflowClassName || "Workflow",
    vellumApiKey: "<TEST_API_KEY>",
  });
}
