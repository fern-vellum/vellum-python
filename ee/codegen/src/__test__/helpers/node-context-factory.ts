import { workflowContextFactory } from "./workflow-context-factory";

import { createNodeContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  SearchNode,
  WorkflowDataNode,
  WorkflowNodeType as WorkflowNodeTypeEnum,
} from "src/types/vellum";

export async function nodeContextFactory({
  workflowContext,
  nodeData,
}: Partial<BaseNodeContext.Args<WorkflowDataNode>> = {}): Promise<
  BaseNodeContext<WorkflowDataNode>
> {
  return await createNodeContext({
    workflowContext: workflowContext ?? workflowContextFactory(),
    nodeData: nodeData ?? {
      id: "search",
      type: WorkflowNodeTypeEnum.SEARCH,
      inputs: [],
      data: { label: "MyNode" } as SearchNode["data"],
    },
  });
}
