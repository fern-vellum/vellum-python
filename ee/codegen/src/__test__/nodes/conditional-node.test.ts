import { Writer } from "@fern-api/python-ast/core/Writer";
import { beforeEach } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import { conditionalNodeFactory } from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { ConditionalNodeContext } from "src/context/node-context/conditional-node";
import { ConditionalNode } from "src/generators/nodes/conditional-node";

describe("ConditionalNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: ConditionalNode;

  beforeEach(async () => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    const nodeData = conditionalNodeFactory();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "d2287fee-98fb-421c-9464-e54d8f70f046",
          key: "field",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as ConditionalNodeContext;
    workflowContext.addNodeContext(nodeContext);

    node = new ConditionalNode({
      workflowContext,
      nodeContext,
    });
  });

  it("getNodeFile", async () => {
    node.getNodeFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("getNodeDisplayFile", async () => {
    node.getNodeDisplayFile().write(writer);
    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
