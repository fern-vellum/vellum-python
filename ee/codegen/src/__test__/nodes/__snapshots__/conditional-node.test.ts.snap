// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConditionalNode > getNodeDisplayFile 1`] = `
"from vellum_ee.workflows.display.nodes import BaseConditionalNodeDisplay
from ...nodes.conditional_node import ConditionalNode
from uuid import UUID
from vellum_ee.workflows.display.nodes.vellum.conditional_node import RuleIdMap
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition


class ConditionalNodeDisplay(BaseConditionalNodeDisplay[ConditionalNode]):
    source_handle_ids = {
        0: UUID("63345ab5-1a4d-48a1-ad33-91bec41f92a5"),
        1: UUID("14a8b603-6039-4491-92d4-868a4dae4c15"),
    }
    rule_ids = [
        RuleIdMap(
            id=UUID("fa50fb0c-8d62-40e3-bd88-080b52efd4b2"),
            lhs=RuleIdMap(
                id=UUID("ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc"), lhs=None, rhs=None
            ),
            rhs=None,
        )
    ]
    condition_ids = {
        0: UUID("8d0d8b56-6c17-4684-9f16-45dd6ce23060"),
        1: UUID("ea63ccd5-3fe3-4371-ba3c-6d3ec7ca2b60"),
    }
    node_input_ids_by_name = {
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.field": UUID(
            "2cb6582e-c329-4952-8598-097830b766c7"
        ),
        "ad6bcb67-f21b-4af9-8d4b-ac8d3ba297cc.value": UUID(
            "cf63d0ad-5e52-4031-a29f-922e7004cdd8"
        ),
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=2247.2797390213086, y=30.917121251477084),
        width=480,
        height=180,
    )
"
`;

exports[`ConditionalNode > getNodeFile 1`] = `
"from vellum.workflows.nodes.displayable import ConditionalNode as BaseConditionalNode
from vellum.workflows.ports import Port
from ..inputs import Inputs


class ConditionalNode(BaseConditionalNode):
    class Ports(BaseConditionalNode.Ports):
        branch_1 = Port.on_if(Inputs.field.equals("testtest"))
        branch_2 = Port.on_else()
"
`;
